protocol_driver_double_barrel.cc

```diff
@@ -0,0 +1,129 @@
+    ProtocolDriverOptions opts, int* port, int tree_depth = 0);
+
+// ProtocolDriver ===============================================
+ProtocolDriverDoubleBarrel::ProtocolDriverDoubleBarrel(int tree_depth) {
+  tree_depth_ = tree_depth;
+}
+
+absl::Status ProtocolDriverDoubleBarrel::Initialize(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+
+  auto pdo = pd_opts;
+  std::string subordinate_driver_type = "";
+  google::protobuf::internal::RepeatedPtrIterator
+      <distbench::NamedSetting> iter_subordinate;
+  for (auto iter=pdo.server_settings().begin();
+       iter!=pdo.server_settings().end();
+       iter++) {
+    if (iter->name()=="next_protocol_driver") {
+      subordinate_driver_type = iter->string_value();
+      break;
+    }
+  }
+  pdo.set_protocol_name(subordinate_driver_type);
+
+  auto maybe_barrel_1 = AllocateProtocolDriver(pdo, &port_1_, tree_depth_+1);
+  if (!maybe_barrel_1.ok()) return maybe_barrel_1.status();
+  barrel_1_ = std::move(maybe_barrel_1.value());
+
+  auto maybe_barrel_2 = AllocateProtocolDriver(pdo, &port_2_, tree_depth_+1);
+  if (!maybe_barrel_2.ok()) return maybe_barrel_2.status();
+  barrel_2_ = std::move(maybe_barrel_2.value());
+
+  return absl::OkStatus();
+}
+
+absl::Status ProtocolDriverDoubleBarrel::InitializeClient(
+    const ProtocolDriverOptions& pd_opts) {
+  return absl::UnimplementedError("InitializeClient is not implemented for double_barrel.");
+}
+
+absl::Status ProtocolDriverDoubleBarrel::InitializeServer(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+  return absl::UnimplementedError("InitializeServer is not implemented for double_barrel.");
+}
+
+void ProtocolDriverDoubleBarrel::SetHandler(
+    std::function<std::function<void()>(ServerRpcState* state)> handler) {
+  barrel_1_->SetHandler(handler);
+  barrel_2_->SetHandler(handler);
+}
+
+void ProtocolDriverDoubleBarrel::SetNumPeers(int num_peers) {
+  barrel_1_->SetNumPeers(num_peers);
+  barrel_2_->SetNumPeers(num_peers);
+}
+
+ProtocolDriverDoubleBarrel::~ProtocolDriverDoubleBarrel() {
+}
+
+absl::StatusOr<std::string> ProtocolDriverDoubleBarrel::HandlePreConnect(
+    std::string_view remote_connection_info, int peer) {
+  return barrel_1_->HandlePreConnect(remote_connection_info, peer);
+  // barrell_2_'s server is not used.
+}
+
+absl::Status ProtocolDriverDoubleBarrel::HandleConnect(
+    std::string remote_connection_info, int peer) {
+  auto ret = barrel_1_->HandleConnect(remote_connection_info, peer);
+  if (!ret.ok()) return ret;
+  return barrel_2_->HandleConnect(remote_connection_info, peer);
+}
+
+void ProtocolDriverDoubleBarrel::HandleConnectFailure(
+    std::string_view local_connection_info) {
+  barrel_1_->HandleConnectFailure(local_connection_info);
+  barrel_2_->HandleConnectFailure(local_connection_info);
+}
+
+std::vector<TransportStat> ProtocolDriverDoubleBarrel::GetTransportStats() {
+  std::vector<TransportStat> stats;
+  return stats;
```

Comment:
call the lower drivers, return what they return, prefix with barrel1/barrel2 or something like that.

Reply:
Yes, but I've kept this for the TextFixture Protocol driver implementation. While exporting TransportStats there, I will have to implement it here as well to work.

Reply:
Please add implementations of all the functions, otherwise this commit will set a bad example for others.

Reply:
Ok Ill add this